controllers/clustertopology_controller.go

```diff
@@ -0,0 +1,177 @@
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusterclasses;machinedeployments,verbs=get;list;watch
+// +kubebuilder:rbac:groups=apiextensions.k8s.io,resources=customresourcedefinitions,verbs=get;list;watch
+
+// ClusterTopologyReconciler reconciles a managed topology for a Cluster object.
+type ClusterTopologyReconciler struct {
+	Client           client.Client
+	WatchFilterValue string
+
+	// UnstructuredCachingClient provides a client that forces caching of unstructured objects,
+	// thus allowing to optimize reads for templates or provider specific objects in a managed topology.
+	UnstructuredCachingClient client.Client
+
+	restConfig      *rest.Config
+	externalTracker external.ObjectTracker
+}
+
+func (r *ClusterTopologyReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	controller, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.Cluster{}).
+		Watches(
+			&source.Kind{Type: &clusterv1.ClusterClass{}},
+			handler.EnqueueRequestsFromMapFunc(r.clusterClassToCluster),
+		).
+		Watches(
+			&source.Kind{Type: &clusterv1.MachineDeployment{}},
+			handler.EnqueueRequestsFromMapFunc(r.machineDeploymentToCluster),
+		).
+		WithOptions(options).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
+		Build(r)
+
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	r.restConfig = mgr.GetConfig()
+	r.externalTracker = external.ObjectTracker{
+		Controller: controller,
+	}
+	return nil
+}
+
+func (r *ClusterTopologyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the Cluster instance.
+	cluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Return early, if the Cluster does not use a managed topology.
+	// NOTE: This should be removed as soon as we start to support Clusters moving from managed <-> unmanaged.
+	if cluster.Spec.Topology == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Return early if the Cluster is paused.
+	if annotations.IsPaused(cluster, cluster) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// TODO: Add patching as soon as we define how to report managed topology state into conditions
+
+	// In case the object is deleted, the managed topology stops to reconcile;
+	// (the other controllers will take care of deletion).
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
The ClusterController waits for dependant objects to be deleted; dependant object deletion is managed by other controllers (KCP, MachineDeployment, MachineSet, Machine controllers) using a combination of finalisers/explicit delete calls or cascading delete via owner dependencies to ensure everything is cleaned up in the provider infrastructure first, then in the API server. 
In case of machined deployment templates, the current implementation is using cascading delete.

Reply:
Also in any case this check should be negative `if !cluster.ObjectMeta.DeletionTimestamp.IsZero() {`

Reply:
Fixed, thanks

Reply:
Got it. Just wasn't sure if (e.g.) KubeadmConfigTemplate and InfrastructureMachineTemplate are considered dependent objects of a MachineDeployment and thus also deleted (I guess because they could be used in theory in multiple clusters)